Add complex solution flows for all languages

datalogics-tsmith opened the pull request:
Up to this point all of our samples have demonstrated using a single endpoint, but the API can also chain endpoints together and we should have samples to demonstrate that.

datalogics-tsmith committed:
Complex flow examples for cURL using multipart requests

datalogics-tsmith committed:
Complex flow examples for DotNET using multipart requests

datalogics-tsmith committed:
Complex flow examples for Javascript using multipart requests

datalogics-tsmith committed:
Complex flow examples for PHP using multipart requests

datalogics-tsmith committed:
Complex flow examples for Python using multipart requests

datalogics-cgreen commented on DotNET/Complex Flow Examples/decrypt-add-reencrypt.cs:
```diff
@@ -0,0 +1,98 @@
+* to modification, running the decrypted result through /pdf-with-added-image,
+* and then sending the output with the new image through /encrypted-pdf to
+* lock it up again.
+*/
```
```suggestion
/* In this sample, we will show how to take an encrypted file and decrypt, modify
* and re-encrypt it to create an encryption-at-rest solution as discussed in
* https://pdfrest.com/solutions/create-secure-document-workflows-with-pdf-password-protection/
* We will be running the document through /decrypted-pdf to open the document
* to modification, running the decrypted result through /pdf-with-added-image,
* and then sending the output with the new image through /encrypted-pdf to
* lock it up again.
*/
```

Just tossed a comma in

datalogics-cgreen commented on DotNET/Complex Flow Examples/merge-different-file-types.cs:
```diff
@@ -0,0 +1,96 @@
+* Note that there is nothing special about an image and a powepoint file and
+* this sample could be easily used to convert and combine any two file types
+* that the /pdf route takes as inputs
+*/
```
I propose we break up the final sentence of the first chunk to improve readability:

```suggestion
/* In this sample, we will show how to merge different file types together as
* discussed in https://pdfrest.com/solutions/merge-multiple-types-of-files-together/.
* First, we will upload an image file to the /pdf route and capture the output ID.
* Next, we will upload a PowerPoint file to the /pdf route and capture its output
* ID. Finally, we will pass both IDs to the /merged-pdf route to combine both inputs
* into a single PDF.

* Note that there is nothing special about an image and a PowerPoint file, and
* this sample could be easily used to convert and combine any two file types
* that the /pdf route takes as inputs.
*/
```
Also added a bit more punctuation.

datalogics-cgreen commented on DotNET/Complex Flow Examples/pdfa-3b-with-attachment.cs:
```diff
@@ -0,0 +1,65 @@
+
+* Note that there is nothing special about attaching an xml file, and any approprite
+* file may be attached and wrapped into the PDF/A conversion.
+*/
```
Minor proofreading changes

```suggestion
/* In this sample, we will show how to attach an xml document to a PDF file and then
* convert the file with the attachment to conform to the PDF/A standard, which
* can be useful for invoicing and standards compliance. We will be running the
* input document through /pdf-with-added-attachment to add the attachment and
* then /pdfa to do the PDF/A conversion.

* Note that there is nothing special about attaching an xml file, and any appropriate
* file may be attached and wrapped into the PDF/A conversion.
*/
```

datalogics-cgreen commented on DotNET/Complex Flow Examples/preserve-word-document.cs:
```diff
@@ -0,0 +1,57 @@
+* We will take our word (or excel or powerpoint) document and first convert it to
+* a PDF with a call to the /pdf route. Then we will take that converted PDF
+* and convert it to the PDF/A format for long term storage.
+*/
```
Spelling, minor grammatical gripes
```suggestion
/* In this sample, we will show how to optimize a Word file for long-term preservation
* as discussed in https://pdfrest.com/solutions/optimize-word-excel-and-powerpoint-files-for-long-term-preservation/
* We will take our Word (or Excel or PowerPoint) document and first convert it to
* a PDF with a call to the /pdf route. Then, we will take that converted PDF
* and convert it to the PDF/A format for long-term storage.
*/
```

datalogics-cgreen commented on DotNET/Complex Flow Examples/protected-watermark.cs:
```diff
@@ -0,0 +1,67 @@
+* https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
+* We will be running the input file through watermarked-pdf to apply the watermark
+* and then /restricted-pdf to lock the watermark in.
+*/
```
```suggestion
/* In this sample, we will show how to watermark a PDF document and then restrict
* editing on the document so that the watermark cannot be removed, as discussed in
* https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
* We will be running the input file through /watermarked-pdf to apply the watermark
* and then /restricted-pdf to lock the watermark in.
*/
```

datalogics-cgreen commented on JavaScript/Complex Flow Examples/merge-different-file-types.js:
```diff
@@ -0,0 +1,87 @@
+* Note that there is nothing special about an image and a powepoint file and
+* this sample could be easily used to convert and combine any two file types
+* that the /pdf route takes as inputs
+*/
```
```suggestion
/* In this sample, we will show how to merge different file types together as
* discussed in https://pdfrest.com/solutions/merge-multiple-types-of-files-together/.
* First, we will upload an image file to the /pdf route and capture the output ID.
* Next, we will upload a PowerPoint file to the /pdf route and capture its output
* ID. Finally, we will pass both IDs to the /merged-pdf route to combine both inputs
* into a single PDF.

* Note that there is nothing special about an image and a PowerPoint file, and
* this sample could be easily used to convert and combine any two file types
* that the /pdf route takes as inputs.
*/
```

datalogics-cgreen commented on PHP/Complex Flow Examples/merge-different-file-types.php:
```diff
@@ -0,0 +1,103 @@
+* Note that there is nothing special about an image and a powepoint file and
+* this sample could be easily used to convert and combine any two file types
+* that the /pdf route takes as inputs
+*/
```
```suggestion
/* In this sample, we will show how to merge different file types together as
* discussed in https://pdfrest.com/solutions/merge-multiple-types-of-files-together/.
* First, we will upload an image file to the /pdf route and capture the output ID.
* Next, we will upload a PowerPoint file to the /pdf route and capture its output
* ID. Finally, we will pass both IDs to the /merged-pdf route to combine both inputs
* into a single PDF.

* Note that there is nothing special about an image and a PowerPoint file, and
* this sample could be easily used to convert and combine any two file types
* that the /pdf route takes as inputs.
*/
```

datalogics-cgreen commented on Python/Complex Flow Examples/merge-different-file-types.py:
```diff
@@ -0,0 +1,97 @@
+
+# Note that there is nothing special about an image and a powepoint file and
+# this sample could be easily used to convert and combine any two file types
+# that the /pdf route takes as inputs
```
```suggestion
# In this sample, we will show how to merge different file types together as
# discussed in https://pdfrest.com/solutions/merge-multiple-types-of-files-together/.
# First, we will upload an image file to the /pdf route and capture the output ID.
# Next, we will upload a PowerPoint file to the /pdf route and capture its output
# ID. Finally, we will pass both IDs to the /merged-pdf route to combine both inputs
# into a single PDF.

# Note that there is nothing special about an image and a PowerPoint file, and
# this sample could be easily used to convert and combine any two file types
# that the /pdf route takes as inputs.
```

datalogics-cgreen commented on cURL/Complex Flow Examples/merge-different-file-types.sh:
```diff
@@ -0,0 +1,40 @@
+
+# Note that there is nothing special about an image and a powepoint file and
+# this sample could be easily used to convert and combine any two file types
+# that the /pdf route takes as inputs
```
```suggestion
# In this sample, we will show how to merge different file types together as
# discussed in https://pdfrest.com/solutions/merge-multiple-types-of-files-together/.
# First, we will upload an image file to the /pdf route and capture the output ID.
# Next, we will upload a PowerPoint file to the /pdf route and capture its output
# ID. Finally, we will pass both IDs to the /merged-pdf route to combine both inputs
# into a single PDF.

# Note that there is nothing special about an image and a PowerPoint file, and
# this sample could be easily used to convert and combine any two file types
# that the /pdf route takes as inputs.
```

datalogics-cgreen commented on JavaScript/Complex Flow Examples/decrypt-add-reencrypt.js:
```diff
@@ -0,0 +1,81 @@
+var FormData = require("form-data");
+var fs = require("fs");
+
+/* In this sample we will show how to take an encrypted file and decrypt, modify
```
```suggestion
/* In this sample, we will show how to take an encrypted file and decrypt, modify
```

datalogics-cgreen commented on PHP/Complex Flow Examples/decrypt-add-reencrypt.php:
```diff
@@ -0,0 +1,112 @@
+use GuzzleHttp\Psr7\Request;
+use GuzzleHttp\Psr7\Utils;
+
+/* In this sample we will show how to take an encrypted file and decrypt, modify
```
```suggestion
/* In this sample, we will show how to take an encrypted file and decrypt, modify
```

datalogics-cgreen commented on Python/Complex Flow Examples/decrypt-add-reencrypt.py:
```diff
@@ -0,0 +1,96 @@
+import requests
+import json
+
+# In this sample we will show how to take an encrypted file and decrypt, modify
```
```suggestion
# In this sample, we will show how to take an encrypted file and decrypt, modify
```

datalogics-cgreen commented on cURL/Complex Flow Examples/decrypt-add-reencrypt.sh:
```diff
@@ -0,0 +1,40 @@
+#!/bin/sh
+
+# In this sample we will show how to take an encrypted file and decrypt, modify
```
```suggestion
# In this sample, we will show how to take an encrypted file and decrypt, modify
```

datalogics-cgreen commented on JavaScript/Complex Flow Examples/pdfa-3b-with-attachment.js:
```diff
@@ -0,0 +1,64 @@
+
+* Note that there is nothing special about attaching an xml file, and any approprite
+* file may be attached and wrapped into the PDF/A conversion.
+*/
```
```suggestion
/* In this sample, we will show how to attach an xml document to a PDF file and then
* convert the file with the attachment to conform to the PDF/A standard, which
* can be useful for invoicing and standards compliance. We will be running the
* input document through /pdf-with-added-attachment to add the attachment and
* then /pdfa to do the PDF/A conversion.

* Note that there is nothing special about attaching an xml file, and any appropriate
* file may be attached and wrapped into the PDF/A conversion.
*/
```

datalogics-cgreen commented on PHP/Complex Flow Examples/pdfa-3b-with-attachment.php:
```diff
@@ -0,0 +1,76 @@
+
+* Note that there is nothing special about attaching an xml file, and any approprite
+* file may be attached and wrapped into the PDF/A conversion.
+*/
```
```suggestion
/* In this sample, we will show how to attach an xml document to a PDF file and then
* convert the file with the attachment to conform to the PDF/A standard, which
* can be useful for invoicing and standards compliance. We will be running the
* input document through /pdf-with-added-attachment to add the attachment and
* then /pdfa to do the PDF/A conversion.

* Note that there is nothing special about attaching an xml file, and any appropriate
* file may be attached and wrapped into the PDF/A conversion.
*/
```

datalogics-cgreen commented on Python/Complex Flow Examples/pdfa-3b-with-attachment.py:
```diff
@@ -0,0 +1,69 @@
+# then /pdfa to do the PDF/A conversion.
+
+# Note that there is nothing special about attaching an xml file, and any approprite
+# file may be attached and wrapped into the PDF/A conversion.
```
```suggestion
# In this sample, we will show how to attach an xml document to a PDF file and then
# convert the file with the attachment to conform to the PDF/A standard, which
# can be useful for invoicing and standards compliance. We will be running the
# input document through /pdf-with-added-attachment to add the attachment and
# then /pdfa to do the PDF/A conversion.

# Note that there is nothing special about attaching an xml file, and any appropriate
# file may be attached and wrapped into the PDF/A conversion.
```

datalogics-cgreen commented on cURL/Complex Flow Examples/pdfa-3b-with-attachment.sh:
```diff
@@ -0,0 +1,27 @@
+# then /pdfa to do the PDF/A conversion.
+
+# Note that there is nothing special about attaching an xml file, and any approprite
+# file may be attached and wrapped into the PDF/A conversion.
```
```suggestion
# In this sample, we will show how to attach an xml document to a PDF file and then
# convert the file with the attachment to conform to the PDF/A standard, which
# can be useful for invoicing and standards compliance. We will be running the
# input document through /pdf-with-added-attachment to add the attachment and
# then /pdfa to do the PDF/A conversion.

# Note that there is nothing special about attaching an xml file, and any appropriate
# file may be attached and wrapped into the PDF/A conversion.
```

datalogics-cgreen commented on JavaScript/Complex Flow Examples/preserve-word-document.js:
```diff
@@ -0,0 +1,56 @@
+ * We will take our word (or excel or powerpoint) document and first convert it to
+ * a PDF with a call to the /pdf route. Then we will take that converted PDF
+ * and convert it to the PDF/A format for long term storage.
+ */
```
```suggestion
/* In this sample, we will show how to optimize a Word file for long-term preservation
* as discussed in https://pdfrest.com/solutions/optimize-word-excel-and-powerpoint-files-for-long-term-preservation/
* We will take our Word (or Excel or PowerPoint) document and first convert it to
* a PDF with a call to the /pdf route. Then, we will take that converted PDF
* and convert it to the PDF/A format for long-term storage.
*/
```

datalogics-cgreen commented on PHP/Complex Flow Examples/preserve-word-document.php:
```diff
@@ -0,0 +1,65 @@
+* We will take our word (or excel or powerpoint) document and first convert it to
+* a PDF with a call to the /pdf route. Then we will take that converted PDF
+* and convert it to the PDF/A format for long term storage.
+*/
```
```suggestion
/* In this sample, we will show how to optimize a Word file for long-term preservation
* as discussed in https://pdfrest.com/solutions/optimize-word-excel-and-powerpoint-files-for-long-term-preservation/
* We will take our Word (or Excel or PowerPoint) document and first convert it to
* a PDF with a call to the /pdf route. Then, we will take that converted PDF
* and convert it to the PDF/A format for long-term storage.
*/
```

datalogics-cgreen commented on Python/Complex Flow Examples/preserve-word-document.py:
```diff
@@ -0,0 +1,66 @@
+# as discussed in https://pdfrest.com/solutions/optimize-word-excel-and-powerpoint-files-for-long-term-preservation/
+# We will take our word (or excel or powerpoint) document and first convert it to
+# a PDF with a call to the /pdf route. Then we will take that converted PDF
+# and convert it to the PDF/A format for long term storage.
```
```suggestion
# In this sample, we will show how to optimize a Word file for long-term preservation
# as discussed in https://pdfrest.com/solutions/optimize-word-excel-and-powerpoint-files-for-long-term-preservation/
# We will take our Word (or Excel or PowerPoint) document and first convert it to
# a PDF with a call to the /pdf route. Then, we will take that converted PDF
# and convert it to the PDF/A format for long-term storage.
```

datalogics-cgreen commented on cURL/Complex Flow Examples/preserve-word-document.sh:
```diff
@@ -0,0 +1,22 @@
+# as discussed in https://pdfrest.com/solutions/optimize-word-excel-and-powerpoint-files-for-long-term-preservation/
+# We will take our word (or excel or powerpoint) document and first convert it to
+# a PDF with a call to the /pdf route. Then we will take that converted PDF
+# and convert it to the PDF/A format for long term storage.
```
```suggestion
# In this sample, we will show how to optimize a Word file for long-term preservation
# as discussed in https://pdfrest.com/solutions/optimize-word-excel-and-powerpoint-files-for-long-term-preservation/
# We will take our Word (or Excel or PowerPoint) document and first convert it to
# a PDF with a call to the /pdf route. Then, we will take that converted PDF
# and convert it to the PDF/A format for long-term storage.
```

datalogics-cgreen commented on JavaScript/Complex Flow Examples/protected-watermark.js:
```diff
@@ -0,0 +1,59 @@
+ * https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
+ * We will be running the input file through watermarked-pdf to apply the watermark
+ * and then /restricted-pdf to lock the watermark in.
+ */
```
```suggestion
/* In this sample, we will show how to watermark a PDF document and then restrict
* editing on the document so that the watermark cannot be removed, as discussed in
* https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
* We will be running the input file through /watermarked-pdf to apply the watermark
* and then /restricted-pdf to lock the watermark in.
*/
```

datalogics-cgreen commented on PHP/Complex Flow Examples/protected-watermark.php:
```diff
@@ -0,0 +1,79 @@
+* https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
+* We will be running the input file through watermarked-pdf to apply the watermark
+* and then /restricted-pdf to lock the watermark in.
+*/
```
```suggestion
/* In this sample, we will show how to watermark a PDF document and then restrict
* editing on the document so that the watermark cannot be removed, as discussed in
* https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
* We will be running the input file through /watermarked-pdf to apply the watermark
* and then /restricted-pdf to lock the watermark in.
*/
```

datalogics-cgreen commented on Python/Complex Flow Examples/protected-watermark.py:
```diff
@@ -0,0 +1,69 @@
+# editing on the document so that the watermark cannot be removed as discussed in
+# https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
+# We will be running the input file through watermarked-pdf to apply the watermark
+# and then /restricted-pdf to lock the watermark in.
```
```suggestion
# In this sample, we will show how to watermark a PDF document and then restrict
# editing on the document so that the watermark cannot be removed, as discussed in
# https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
# We will be running the input file through /watermarked-pdf to apply the watermark
# and then /restricted-pdf to lock the watermark in.
```

datalogics-cgreen commented on cURL/Complex Flow Examples/protected-watermark.sh:
```diff
@@ -0,0 +1,27 @@
+# editing on the document so that the watermark cannot be removed as discussed in
+# https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
+# We will be running the input file through watermarked-pdf to apply the watermark
+# and then /restricted-pdf to lock the watermark in.
```
```suggestion
# In this sample, we will show how to watermark a PDF document and then restrict
# editing on the document so that the watermark cannot be removed, as discussed in
# https://pdfrest.com/solutions/add-pdf-watermarks-that-cannot-be-removed/.
# We will be running the input file through /watermarked-pdf to apply the watermark
# and then /restricted-pdf to lock the watermark in.
```

datalogics-cgreen commented on DotNET/Complex Flow Examples/decrypt-add-reencrypt.cs:
```diff
@@ -0,0 +1,98 @@
+
+using (var httpClient = new HttpClient { BaseAddress = new Uri("https://api.pdfrest.com") })
+{
+    using (var decryptRequest = new HttpRequestMessage(HttpMethod.Post, "decrypted-pdf"))
```
All but the topmost `using`s can simplified in order to avoid a cascade of additional curly braces. Example:

```
using var decryptRequest = new HttpRequestMessage(HttpMethod.Post, "decrypted-pdf");
```

Consequently (or maybe in any case), it may help to add a small comment marking the start of each step outlined by the topmost comment.

datalogics-cgreen commented on DotNET/Complex Flow Examples/decrypt-add-reencrypt.cs:
```diff
@@ -0,0 +1,98 @@
+{
+    using (var decryptRequest = new HttpRequestMessage(HttpMethod.Post, "decrypted-pdf"))
+    {
+        decryptRequest.Headers.TryAddWithoutValidation("Api-Key", "xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx");
```
Generally, we ought to store the API key in a `const` variable or some other, single location so you don't have to replace multiple instances of `xxxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` in the same file.

datalogics-cgreen reviewed (changes requested):
This is an initial, pretty much visual pass-through.

Most of the suggested changes are essentially the same: they concern the same five inline comments on each set of sample files for each language. Should you accept a suggested change, you should be able to bundle all of the related suggestions into a single commit via GitHub ("Add suggestion to batch").

datalogics-tsmith committed:
Grammar changes for the complex flow samples

datalogics-tsmith committed:
Add key variable for complex calls

So that the end user does not have to replace the key in multiple locations
in order to run the sample

datalogics-tsmith committed:
Remove extra 'using' statements from dotnet samples

datalogics-cgreen reviewed (changes requested):
re: [29a52ea](https://github.com/datalogics/pdf-rest-api-samples/pull/61/commits/29a52ea594a127b35698159651139931216f05c8)

This wasn't precisely what I suggested -- simplifying the `using` statements entails simply capping them off with `;` rather than starting a new `{ }` block.

Keeping them in in some capacity would set a better example IMO, so I'm cool with either reverting/resetting this commit (faster), or reintroducing `using` to these lines.

datalogics-tsmith committed:
Re-add using without brackets

For best practices

datalogics-cgreen reviewed (approved):
Thanks! LGTM
